File: problem sets/00/sample code/Java00/ModuleExamples.java
// Examples for Problem Set 00.
//
// Students should write their own examples for complete glass-box
// testing of their implementations.
//
// The examples used for grading will not be exactly the same as
// these examples.

import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;


/**
 * Defines some examples for testing and makes them available as
 * public members of the class.
 */

public class ModuleExamples {

    // abbreviation

    private static Module makeModule (String name,
                                      List<String> uses,
                                      int mtime,
                                      int ctime,
                                      String cmode) {
        return Modules.makeModule (name, uses, mtime, ctime, cmode);
    }

    // a description of a module that's been modified since it was compiled
    // and must therefore be compiled again before it is used

    public static Module foo
        = makeModule ("Foo", list ("Baz"), 1505109465, 1504097449, "LP64");

    // a list of modules that have a circular dependency

    public static List<Module> circularModules
        = list (makeModule ("M1", list ("M2", "M3"), 1500, -1, "ILP32"),
                makeModule ("M2", list ("M3"),       2000, -1, "ILP32"),
                makeModule ("M3", list ("M2"),       2500, -1, "ILP32"));

    // a list of modules, of which only Main needs to be compiled
    // (in mode LP64) before Main can be used

    public static List<Module> modules1
        = list (makeModule ("Main",
                            list ("List", "AList"),
                            1504188920,
                            -1,
                            "LP64"),
                makeModule ("List",
                            list ("Obj"),
                            1472652920,
                            1472658760,
                            "LP64"),
                makeModule ("AList",
                            list ("List", "Obj"),
                            1472654764,
                            1472659242,
                            "LP64"),
                makeModule ("Obj",
                            list(),
                            1472630256,
                            1472638841,
                            "LP64"));

    // a list of modules, of which List, AList, and Main need to be compiled
    // (in that order, in mode ILP32) before Main can be used

    public static List<Module> modules2
        = list (makeModule ("Main",
                            list ("List", "AList"),
                            1504188920,
                            -1,
                            "LP64"),
                makeModule ("List",
                            list ("Obj"),
                            1472652920,
                            1472658760,
                            "LP64"),
                makeModule ("AList",
                            list ("List", "Obj"),
                            1472654764,
                            1472659242,
                            "LP64"),
                makeModule ("Obj",
                            list(),
                            1472630256,
                            1472638841,
                            "ILP32"));

    // more abbreviations

    public static List<String> list () {
        List<String> result = new ArrayList<String>();
        return result;
    }

    public static List<String> list (String s1) {
        List<String> result = new LinkedList<String>();
        result.add (s1);
        return result;
    }

    public static List<String> list (String s1, String s2) {
        List<String> result = new ArrayList<String>();
        result.add (s1);
        result.add (s2);
        return result;
    }

    public static List<String> list (String s1, String s2, String s3) {
        List<String> result = new LinkedList<String>();
        result.add (s1);
        result.add (s2);
        result.add (s3);
        return result;
    }

    public static List<String> list (String s1,
                                     String s2,
                                     String s3,
                                     String s4) {
        List<String> result = new ArrayList<String>();
        result.add (s1);
        result.add (s2);
        result.add (s3);
        result.add (s4);
        return result;
    }

    public static List<Module> list (Module m1) {
        List<Module> result = new ArrayList<Module>();
        result.add (m1);
        return result;
    }

    public static List<Module> list (Module m1, Module m2) {
        List<Module> result = new LinkedList<Module>();
        result.add (m1);
        result.add (m2);
        return result;
    }

    public static List<Module> list (Module m1, Module m2, Module m3) {
        List<Module> result = new ArrayList<Module>();
        result.add (m1);
        result.add (m2);
        result.add (m3);
        return result;
    }

    public static List<Module> list (Module m1,
                                     Module m2,
                                     Module m3,
                                     Module m4) {
        List<Module> result = new LinkedList<Module>();
        result.add (m1);
        result.add (m2);
        result.add (m3);
        result.add (m4);
        return result;
    }
}

File: problem sets/00/solution/moduleexamples.py


from modules import Modules
from PS00 import PS00


class ModuleExamples:

    # def _makeModule(self, name, uses, mtime, ctime, cmode):
    #     return Modules.makeModule(name, uses, mtime, ctime, cmode)

    # a description of a module that's been modified since it was compiled
    # and must therefore be compiled again before it is used
    foo = Modules.makeModule("Foo", ["Baz"], 1505109465, 1504097449, "LP64")

    # a list of modules that have a circular dependency
    circularModules = [Modules.makeModule("M1", ["M2", "M3"], 1500, -1, "ILP32"),
                       Modules.makeModule("M2", ["M3"], 2000, -1, "ILP32"),
                       Modules.makeModule("M3", ["M2"], 2500, -1, "ILP32")]

    # a list of modules, of which only Main needs to be compiled
    # (in mode LP64) before Main can be used
    modules1 = [Modules.makeModule("Main", ["List", "AList"], 1504188920, -1, "LP64"),
                Modules.makeModule(
                    "List", ["Obj"], 1472652920, 1472658760, "LP64"),
                Modules.makeModule(
                    "AList", ["List", "Obj"], 1472654764, 1472659242, "LP64"),
                Modules.makeModule("Obj", [], 1472630256, 1472638841, "LP64")]

    # a list of modules, of which List, AList, and Main need to be compiled
    # (in that order, in mode ILP32) before Main can be used
    modules2 = [Modules.makeModule("Main", ["List", "AList"], 1504188920, -1, "LP64"),
                Modules.makeModule(
                    "List", ["Obj"], 1472652920, 1472658760, "LP64"),
                Modules.makeModule(
                    "AList", ["List", "Obj"], 1472654764, 1472659242, "LP64"),
                Modules.makeModule("Obj", [], 1472630256, 1472638841, "ILP32")]

    modules3 = [Modules.makeModule("Main", ["List", "AList"], 0, 11, "LP64"),
                Modules.makeModule(
                    "List", ["Obj"], 0, 11, "LP64"),
                Modules.makeModule(
                    "AList", ["List", "Obj"], 0, 11, "LP64"),
                Modules.makeModule("Obj", [], 0, 11, "LP64")]

    modules4 = [Modules.makeModule("Main", ["List", "AList"], 3, 10, "ILP32"),
                Modules.makeModule("List", ["Obj"], 1, 4, "ILP64"),
                Modules.makeModule("AList", ["A1", "A2"], 5, 4, "ILP32"),
                Modules.makeModule("Obj", ["O1", "O2"], 1, 2, "LP32"),
                Modules.makeModule("O1", [], 1, 3, "LP32"),
                Modules.makeModule("O2", ["A3"], 1, 4, "LP32"),
                Modules.makeModule("A1", [], 1, 2, "LP64"),
                Modules.makeModule("A2", ["A3"], 1, 2, "LP64"),
                Modules.makeModule("A3", [], 2, -1, "LP64")]


if __name__ == '__main__':
    soln = PS00()
    print('Circular Test for circularmodules ===>', soln.isCircular(ModuleExamples.circularModules))
    print('Circular Test for modules1 ===>', soln.isCircular(ModuleExamples.modules1))
    print('Circular Test for modules2 ===>', soln.isCircular(ModuleExamples.modules2))
    print('Circular Test for modules3 ===>', soln.isCircular(ModuleExamples.modules3))
    print('Circular Test for modules4 ===>', soln.isCircular(ModuleExamples.modules4))
    print('BestMode Test for modules1 ===>', soln.bestMode('Main', ModuleExamples.modules1))
    print('BestMode Test for modules2 ===>', soln.bestMode('Main', ModuleExamples.modules2))
    print('BestMode Test for modules3 ===>', soln.bestMode('Main', ModuleExamples.modules3))
    print('BestMode Test for modules4 ===>', soln.bestMode('Main', ModuleExamples.modules4))
    print('BestPlan Test for modules1 ===>', soln.bestPlan('Main', ModuleExamples.modules1))
    print('BestPlan Test for modules2 ===>', soln.bestPlan('Main', ModuleExamples.modules2))
    print('BestPlan Test for modules3 ===>', soln.bestPlan('Main', ModuleExamples.modules3))
    print('BestPlan Test for modules4 ===>', soln.bestPlan('Main', ModuleExamples.modules4))
    print('============= FIN =============')
